### ISSUE TITLE
Show precedents that are similar to the current claim

### ISSUE THREAD

**Issue by arekmano** [Cyberjusticelab/JusticeAI#261]
*Description*
As a user, I would like the system to show me the 5 most similar precedents to my claim

*Scope of Work*
- [x] Investigate if there are similarity metrics that are easily calculable given our SVM weights (Probably not, but still worth looking). [Update: nearest neighbour calculation required additional distance calculation, as the SVM's weights are only related to answering a single specific question] 
- [x] Implement 5-NN algorithm using sci-kit learn
- [x] Create UI response that outputs the case number of the precedents that are most similar, as well as a similarity metric (% similar or distance measure)

*Acceptance Criteria*
- The user fills out claim facts
- Chatbot responds with top 5 most similar case numbers

**Comment by arekmano** [comment]
This user story likely involves finding the 3 or 4 nearest neighbors of the data point, under a threshold distance value

**Comment by dodoels** [comment]
Are we allow to show the precedents?

**Comment by Vynny** [comment]
As @choitwao mentioned, are we allowed to show this data?
I'm uncertain that showing users case numbers is going to offer them any value, unless we somehow summarize every case and present that on the front end.

Perhaps just offering a count? 
Example: "I know of 4 cases that are similar to yours"

**Comment by arekmano** [comment]
just display the case numbers as a POC. This we are 100% fine with displaying.

**Comment by arekmano** [comment]
Precedent similarity was implemented using euclidean distance, on the assumption that smaller distance means that the data points are more closely related.

This assumption holds with features that are binary. However, as <a href="https://link.springer.com/chapter/10.1007/3-540-44503-X_27">this paper</a> indicates, this assumption may not always hold true. We will have to reevaluate the approach once our features are converted to values as part of #251.

**Comment by arekmano** [comment]
One possible solution to the problem could be to use Mahalanobi distance, rather than euclidean, which would take into account the variance of each feature, therefore reducing the effect of large values (i.e. monthly dollar costs)

Relevant Paper:
http://www.sciencedirect.com/science/article/pii/S0169743999000477

**Comment by arekmano** [comment]
In order to evaluate the best distance metric, manual precedent evaluation must be performed.

After this, we can determine how accurate the measurement is in detecting similarity

**Comment by arekmano** [comment]
An additional similarity metric that we can use is the Cosine Similarity (which is also supported within SciKit Learn: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html#sklearn-metrics-pairwise-cosine-similarity)

**Comment by Samuel-Campbell** [comment]
Simplified explanation for Mahalanobis distance:
https://en.wikipedia.org/wiki/Mahalanobis_distance

**Comment by naregeff** [comment]
It predicts similarity! Accuracy set to improve as project moves forward.
